Sources/KSCrashRecording/Monitors/KSCrashMonitor_Memory.m

```diff
@@ -0,0 +1,837 @@
+    // and indication of an OOM.
+    bool userPerceivedOOM = NO;
+    if (ksmemory_previous_session_was_terminated_due_to_memory(&userPerceivedOOM)) {
+        
+        // We only report an OOM that the user might have seen.
+        // Ignore this check if we want to report all OOM, foreground and background.
+        if (userPerceivedOOM) {
+            NSURL *url = kscm_memory_oom_breadcrumb_URL();
+            const char *reportContents = kscrash_readReportAtPath(url.path.UTF8String);
+            if (reportContents) {
+                
+                NSData *data = [NSData dataWithBytes:reportContents length:strlen(reportContents)];
+                NSMutableDictionary *json = [[NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers|NSJSONReadingMutableLeaves error:nil] mutableCopy];
+                
+                if (json) {
+                    json[@KSCrashField_System][@KSCrashField_AppMemory] = kscm_memory_serialize(&g_previousSessionMemory);
+                    json[@KSCrashField_Report][@KSCrashField_Timestamp] = @(g_previousSessionMemory.timestamp);
+                    json[@KSCrashField_Crash][@KSCrashField_Error][@KSCrashExcType_MemoryTermination] =kscm_memory_serialize(&g_previousSessionMemory);
+                    json[@KSCrashField_Crash][@KSCrashField_Error][@KSCrashExcType_Mach] = nil;
+                    json[@KSCrashField_Crash][@KSCrashField_Error][@KSCrashExcType_Signal] = @{
+                        @KSCrashField_Signal: @(SIGKILL),
+                        @KSCrashField_Name: @"SIGKILL",
+                    };
+                    
+                    data = [NSJSONSerialization dataWithJSONObject:json options:NSJSONWritingPrettyPrinted error:nil];
+                    kscrash_addUserReport((const char *)data.bytes, (int)data.length);
+                }
+                free((void *)reportContents);
+            }
+        }
+    }
+    
+    // remove the old breadcrumb oom file
+    unlink(kscm_memory_oom_breadcrumb_URL().path.UTF8String);
+}
+
+/**
+ This is called after all monitors are enabled.
+ */
+static void notifyPostSystemEnable(void)
+{
+    if (g_hasPostEnable) {
+        return;
+    }
+    g_hasPostEnable = 1;
+    
+    kscm_memory_check_for_oom_in_previous_session();
+
+    if (g_isEnabled)
+    {
+        ksmemory_write_possible_oom();
+        g_memoryTracker = [[MemoryTracker alloc] init];
+    }
+}
+
+KSCrashMonitorAPI* kscm_memory_getAPI(void)
+{
+    static KSCrashMonitorAPI api =
+    {
+        .setEnabled = setEnabled,
+        .isEnabled = isEnabled,
+        .addContextualInfoToEvent = addContextualInfoToEvent,
+        .notifyPostSystemEnable = notifyPostSystemEnable,
+    };
+    return &api;
+}
+
+/**
+ Read the previous sessions memory data.
+ */
+static void ksmemory_read(const char* path)
+{
+    int fd = open(path, O_RDONLY, 0644);
+    if (fd == -1) {
+        unlink(path);
```

Comment:
Good call on the mode. I'll keep the unlink since it's purpose is not to fix and problems during open but to get rid of any file that might be there (even if open fails).